### ISSUE TITLE
Unexpected result when using CASE operator or transform function with NULL

### ISSUE THREAD

**Issue by siberex** [ClickHouse/ClickHouse#2655]
Expected result:

```sql
SELECT transform(0, [0, 1], ['ZERO', 'ONE'], 'DEFAULT') AS result
-- ZERO
```

Unexpected result, always gives NULL:
```sql
SELECT transform(0, [0, 1], ['ZERO', 'ONE'], NULL) AS result
-- ┌─result─┐
-- │ ᴺᵁᴸᴸ   │
-- └────────┘
--
-- 1 rows in set. Elapsed: 0.003 sec.
```

Same for the `CASE` operator:

```sql
SELECT CASE 1
  WHEN 0 THEN 'ZERO'
  WHEN 1 THEN 'ONE'
  ELSE 'NONE'
END AS result
-- ONE

SELECT CASE 1
  WHEN 0 THEN NULL
  WHEN 1 THEN 'ONE'
  ELSE 'NONE'
END AS result
-- ┌─result─┐
-- │ ᴺᵁᴸᴸ   │
-- └────────┘
```

ClickHouse client version 1.1.54390.
ClickHouse server version 1.1.54390.

**Comment by siberex** [comment]
It is unexpected because it have completely different implementation in, like, literally any other SQL-flavoured DB:

```
mysql> SELECT CASE 1
    ->   WHEN 0 THEN 'ZERO'
    ->   WHEN 1 THEN 'ONE'
    ->   ELSE 'NONE'
    -> END AS result;
+--------+
| result |
+--------+
| ONE    |
+--------+
1 row in set (0.00 sec)

mysql> SELECT CASE 1
    ->   WHEN 0 THEN NULL
    ->   WHEN 1 THEN 'ONE'
    ->   ELSE 'NONE'
    -> END AS result;
+--------+
| result |
+--------+
| ONE    |
+--------+
1 row in set (0.00 sec)
```

**Comment by alexey-milovidov** [comment]
The issue is still relevant. Reference: #9596

**Comment by alexey-milovidov** [comment]
Fixed in 23.5.
